### PR TITLE
FilterOptions: filter by locationRadius w/ optionType RADIUS_FILTER

### PR DESCRIPTION
**Pull request by oshi97**
This commit adds a new config option for FilterOptions called optionType, which
can be either RADIUS_FILTER or STATIC_FILTER, defaulting to STATIC_FILTER which
is the original behavior. Filtering by radius is done by appending a new
locationRadius param to a query.

TEST=manual
Test Facets, FilterBox, FilterOptions
Facets: smoke test
FilterBox:
- check that regular STATIC_FILTER FilterBox still works, and can filter by distance
- check that reseting FilterBox then applying also clears the locationRadius
- check that works with searchOnChange (after fixing reset button init bug with #768 )

FilterOptions:
- check that defaults to STATIC_FILTER, throws helpful errors
when trying to use unrecognized optionType or trying to use multioption
with RADIUS_FILTER
- check that setting a value of 0 for RADIUS_FILTER clears locationRadius
- check that selected filter persists over page refresh
- check that can filter by locationRadius by searching "new york" under vertical key "people"
  locationRadius = 1 gives no results, locationRadius = 8046.72 = 1 result
- check that clicking the reset button in FilterOptions then searching again clears the 
  locationRadius

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1140,6 +1122,74 @@ ANSWERS.addComponent('FilterOptions', {
...
 
+The options config varies depending on whether the optionType is 'STATIC_FILTER' or 'RADIUS_FILTER'.
+
+##### STATIC_FILTER
+
+```js
+{
+  options: [
+    {
+      // Label to show next to the filter option
```
Comment (alexisgrow) [inline README.md]: Can these comments have a "Required" or "Optional" prefix?

Comment (oshi97) [reply README.md]: updated! Reordered things to have required options first

File: README.md [hunk 2]
```diff
@@ -1085,33 +1085,15 @@ ANSWERS.addComponent('FilterOptions', {
   container: '.filter-container',
   // Control type, singleoption or multioption
   control: 'singleoption',
+  // The type of filtering to apply to the options, either 'STATIC_FILTER' or 'RADIUS_FILTER'.
+  // Defaults to 'STATIC_FILTER'.
```
Comment (tmeyer2115) [inline README.md:1061]: I think it might be more accurate to say that optionType determines the kind of the options, not so much the filtering applied to them. STATIC_FILTER type options refer to values for a profile field, right? RADIUS_FILTER options are distances.

File: src/ui/components/filters/filteroptionscomponent.js [hunk 3]
```diff
@@ -141,7 +155,9 @@ class FilterOptionsConfig {
     if (this.control === 'singleoption') {
       const _options = options.map(o => ({ ...o }));
       const firstAppliedOption = _options.find(o => o.appliedOnLoad);
-      firstAppliedOption.selected = true;
+      if (firstAppliedOption) {
+        firstAppliedOption.selected = true;
+      }
       return _options;
```
Comment (tmeyer2115) [inline src/ui/components/filters/filteroptionscomponent.js]: I think this method can be made more efficient. We're essentially traversing the options array twice. Once with the .map and again with the .find. We should be able to consolidate the two traversals by making the mapping function a bit more complex.

**Comment by oshi97** [comment]
rebased applied-filters onto v0.14.0, distance-filter onto applied-filters, then this branch onto distance-filter

File: src/ui/components/filters/filteroptionscomponent.js [hunk 4]
```diff
@@ -322,8 +351,22 @@ export default class FilterOptionsComponent extends Component {
   }
 
   apply () {
-    const filter = this._buildFilter();
-    this.core.setFilter(this.name, filter);
+    switch (this.config.optionType) {
+      case OptionTypes.RADIUS_FILTER:
```
Comment (tmeyer2115) [inline src/ui/components/filters/filteroptionscomponent.js:355]: Minor nit, but could consolidate things a bit here. 
const selectedOption = this.config.options.find(o => o.selected && o.value !== 0);
selectedOption && this.core.setLocationRadius(selectedOption.value);

File: src/ui/components/filters/filteroptionscomponent.js (hunk 3 above)
Comment (oshi97) [reply src/ui/components/filters/filteroptionscomponent.js]: I was worrying about not modifying the original array but it's pretty idiotic to worry about that when the array is just being set to the return value anyways. I don't really want to make the function more complex just to force the code to use .map

File: src/ui/components/filters/filteroptionscomponent.js (hunk 4 above)
Comment (oshi97) [reply src/ui/components/filters/filteroptionscomponent.js:355]: I feel like this is not logically equivalent. If the selected value is === 0 I still want to call core.clearLocationRadius. I could move the "if the value is 0 clear the locationRadius" logic into core.setLocationRadius and delete core.clearLocationRadius though, if you like that better? The backend will throw an error if you set locationRadius to 0 so I think that makes sense, but maybe we don't want to tie our behavior with theirs.

File: README.md (hunk 2 above)
Comment (oshi97) [reply README.md:1061]: I wanted to avoid documentation that basically takes a config option like chocolateChipCookie and tells you "it is a cookie made with chocolate chips" but I think what you said makes sense here, updated!

File: src/ui/components/filters/filteroptionscomponent.js (hunk 3 above)
Comment (oshi97) [reply src/ui/components/filters/filteroptionscomponent.js]: removed .map() call!

**Comment by oshi97** [comment]
The spec was changed from `appliedOnLoad` to `selected` so this PR was changed to match. This was originally going to go in a separate PR but it looks like it would make things simpler for everyone to just put it here.

**Comment by oshi97** [comment]
FilterOptions: rename 'appliedOnLoad' config option to 'selected'

THe 'selected' config option already exists (as a by product of implementation
details) but the spec has been changed to use 'selected' instead of 'appliedOnLoad'
and it doesn't look like there are any issues with this.

TEST=manual
Facets: smoke test
FilterOptions: check that works with multioption/singleoption, only selecting
the first selected option for singleoption, and using persistentStorage instead
if any stored data exists

File: README.md (hunk 2 above)
Comment (tmeyer2115) [reply README.md:1061]: I think it's okay to say that optionType represents the type or kind of the options if we also provide some quick details about where 'STATIC_FILTER' options (profile field values) and 'RADIUS_FILTER' options (distances) come from. If we provide the additional details about what the different kinds of options represent, you're providing the reader with additional info apart from the 'optionType is type of options' tautology. You don't need much detail here since you expand on them below.

File: src/ui/components/filters/filteroptionscomponent.js [hunk 5]
```diff
@@ -134,20 +148,27 @@ class FilterOptionsConfig {
...
   /**
    * If no previous options were stored in persistentStorage, default to options marked
-   * as appliedOnLoad.
+   * as selected. If multiple options are marked as selected for 'singleoption', only the
+   * first should be selected.
    * @param {*} options
+   * @returns {Array<Object>}
    */
-  setAppliedOnLoad (options) {
+  setSelected (options) {
```
Comment (tmeyer2115) [inline src/ui/components/filters/filteroptionscomponent.js]: Can we consolidate setPreviousOptions and setSelected into one, single setSelected method? I think it could help reduce a bit of complexity if we had a single method whose responsibility was setting the selected attribute for an option(s). As it stands, we essentially have two different methods for doing this.

Comment (oshi97) [reply src/ui/components/filters/filteroptionscomponent.js]: updated!

File: src/ui/components/filters/filteroptionscomponent.js [hunk 6]
```diff
@@ -126,35 +126,25 @@ class FilterOptionsConfig {
...
    * first should be selected.
-   * @param {*} options
+   * @param {Array<Object>} options
+   * @param {Array<string>} previousOptions
    * @returns {Array<Object>}
    */
-  setSelected (options) {
-    if (this.control === 'singleoption') {
+  setSelectedOptions (options, previousOptions) {
+    if (previousOptions) {
```
Comment (tmeyer2115) [inline src/ui/components/filters/filteroptionscomponent.js]: I think there's a weird edge case here. Someone could first configure FilterOptions to be 'multioption', then save multiple selected options in persistent storage. If they then decide to change FilterOptions to be 'singleoption', but use the URL from before, could they wind up with multiple options selected here, even though it's 'singleoption'?

Comment (oshi97) [reply src/ui/components/filters/filteroptionscomponent.js]: Ah I didn't consider that. What happens is that multiple radio buttons are marked as checked in the html. In Chrome and Firefox this is interpreted as visually only the last radio button being selected, however if you inspect the dom multiple radio buttons are given the 'checked' attribute. If you apply the filters, since the component doesn't care about whether it's multioption/singleoption when it builds the filter it includes all the selected options.

Is this something we need to handle? I think no matter what we do the url won't be valid anymore.

Comment (tmeyer2115) [reply src/ui/components/filters/filteroptionscomponent.js]: So they could generate a filter with multiple options selected even though it's type 'singleoption'? I think we should handle this, even though it's an extreme edge case. For consistency's sake, I think we should just pick the first of the previousOptions in this case. What do you think?

Comment (oshi97) [reply src/ui/components/filters/filteroptionscomponent.js]: yeah sounds good to me!

Comment (oshi97) [reply src/ui/components/filters/filteroptionscomponent.js]: updated to add this handling, also added unit tests for RADIUS_FILTER and options.selected logic

File: src/ui/components/filters/filteroptionscomponent.js [hunk 7]
```diff
@@ -25,11 +33,17 @@ class FilterOptionsConfig {
...
+     * @type {string}
+     */
+    this.optionType = config.optionType || OptionTypes.STATIC_FILTER;
+
     /**
      * The list of filter options to display with checked status
      * @type {object[]}
      */
-    this.options = config.options;
+    this.options = config.options.map(o => ({ ...o }));
```
Comment (tmeyer2115) [inline src/ui/components/filters/filteroptionscomponent.js:46]: Minor nit, you can use slice() to make a copy

File: tests/ui/components/filters/filteroptionscomponent.js [hunk 8]
```diff
@@ -233,3 +229,267 @@ describe('filter options component', () => {
     expect(setStaticFilterNodes.mock.calls).toHaveLength(6);
   });
 });
+
+describe('filter options component selected options', () => {
```
Comment (tmeyer2115) [inline tests/ui/components/filters/filteroptionscomponent.js]: Can you provide a more descriptive name for this suite of tests? The name doesn't make it very clear what's being tested.

File: tests/ui/components/filters/filteroptionscomponent.js [hunk 9]
```diff
@@ -233,3 +229,267 @@ describe('filter options component', () => {
...
+      ]
+    };
+
+    const component = COMPONENT_MANAGER.create('FilterOptions', config);
+    const options = component.config.options;
+    expect(options).toHaveLength(3);
+    expect(options.filter(o => o.selected)).toHaveLength(0);
+  });
+
+  it('can choose selected options for multioption', () => {
```
Comment (tmeyer2115) [inline tests/ui/components/filters/filteroptionscomponent.js]: How about 'properly sets selected options for multioption'? This kinda makes it sound like we're testing that a user can click multiple options.

File: tests/ui/components/filters/filteroptionscomponent.js [hunk 10]
```diff
@@ -233,3 +229,267 @@ describe('filter options component', () => {
...
+    const component = COMPONENT_MANAGER.create('FilterOptions', config);
+    const options = component.config.options;
+    expect(options).toHaveLength(4);
+    const selectedOptions = options.filter(o => o.selected);
+    expect(selectedOptions).toHaveLength(2);
+    expect(selectedOptions[0].label).toEqual('label3');
+    expect(selectedOptions[1].label).toEqual('label4');
+  });
+
+  it('prioritizes previous options over selected config for multioption', () => {
```
Comment (tmeyer2115) [inline tests/ui/components/filters/filteroptionscomponent.js]: 'prioritizes previously selected options over selected options in config for multioption'

File: tests/ui/components/filters/filteroptionscomponent.js [hunk 11]
```diff
@@ -233,3 +229,267 @@ describe('filter options component', () => {
...
+    const component = COMPONENT_MANAGER.create('FilterOptions', config);
+    const options = component.config.options;
+    expect(options).toHaveLength(4);
+    const selectedOptions = options.filter(o => o.selected);
+    expect(selectedOptions).toHaveLength(2);
+    expect(selectedOptions[0].label).toEqual('label1');
+    expect(selectedOptions[1].label).toEqual('label2');
+  });
+
+  it('can choose selected option for singleoption', () => {
```
Comment (tmeyer2115) [inline tests/ui/components/filters/filteroptionscomponent.js]: 'properly sets selected option for singleoption'

File: tests/ui/components/filters/filteroptionscomponent.js [hunk 12]
```diff
@@ -233,3 +229,267 @@ describe('filter options component', () => {
...
+
+    const component = COMPONENT_MANAGER.create('FilterOptions', config);
+    const options = component.config.options;
+    expect(options).toHaveLength(4);
+    const selectedOptions = options.filter(o => o.selected);
+    expect(selectedOptions).toHaveLength(1);
+    expect(selectedOptions[0].label).toEqual('label3');
+  });
+
+  it('prioritizes previous options over selected config for singleoption', () => {
```
Comment (tmeyer2115) [inline tests/ui/components/filters/filteroptionscomponent.js]: 'prioritizes previously selected option over selected option in config for singleoption'

File: src/ui/components/filters/filteroptionscomponent.js (hunk 7 above)
Comment (oshi97) [reply src/ui/components/filters/filteroptionscomponent.js:46]: for the record I did NOT know about this tip, but after some testing it looks like slice doesn't make a deep copy of the objects. What I did was
```
let a = [{ a: 123}, {a: 345, b: true}]
let b = a.slice()
b[1].a = 123
console.log(a[1].a === b[1].a)
```

and in this case array `a` was also modified

File: tests/ui/components/filters/filteroptionscomponent.js (hunk 8 above)
Comment (oshi97) [reply tests/ui/components/filters/filteroptionscomponent.js]: updated!

File: tests/ui/components/filters/filteroptionscomponent.js (hunk 9 above)
Comment (oshi97) [reply tests/ui/components/filters/filteroptionscomponent.js]: thank you for your brainpower

File: src/core/filters/filterregistry.js [hunk 13]
```diff
@@ -76,6 +76,16 @@ export default class FilterRegistry {
     return Facet.fromFilters(this.availableFieldIds, ...filters);
   }
 
+  /**
+   * Gets the locationRadius to send in a search query.
+   * @returns {number}
+   */
+  getLocationRadiusPayload () {
```
Comment (tmeyer2115) [inline src/core/filters/filterregistry.js]: Are all these payload methods used only in core.js? If so, I think we should move them to that class. Arguably, I think it's beyond the responsibility of this class to have payload methods. The role of this class is to maintain a collection of filters. It should have setters for adding filter nodes and getters as well. It deals in the currency of FilterNodes. Consumers of this class can then translate filter nodes into whatever structure they need. 

I'd suggest we make an item to remove the existing payload methods. For this particular method, what if we instead add a getFilterNodeByKey method to the registry? We have getters for all static filters, facets, etc. Might be nice to have a general purpose method for pulling exactly one filter, by key. In core.js, we can use this method to get the location radius FilterNode, then get the enclosed value.

File: src/core/models/filter.js [hunk 14]
```diff
@@ -192,6 +192,16 @@ export default class Filter {
     return Filter._fromMatcher('builtin.location', '$near', { lat, lng, radius });
   }
 
+  /**
+   * Create a new RADIUS_FILTER.
+   * @param {number} radius
+   */
+  static locationRadius (radius) {
```
Comment (tmeyer2115) [inline src/core/models/filter.js]: This method doesn't do much besides call the Filter constructor. Is it really necessary? I'm hesitant to clutter the Filter class with more of these static method. There are already a lot.

File: src/core/filters/filterregistry.js [hunk 15]
```diff
@@ -76,6 +76,16 @@ export default class FilterRegistry {
...
+    const filterNode =
+      this.globalStorage.getState(`${this.filterRegistryId}.${StorageKeys.LOCATION_RADIUS}`);
```
Comment (tmeyer2115) [inline src/core/filters/filterregistry.js]: As we talked about last night, I think we should remove filterRegistryId as it's no longer being used. For the sake of separation of concerns, I think we should do that in a separate PR.

File: src/ui/components/filters/filteroptionscomponent.js [hunk 16]
```diff
@@ -313,6 +386,28 @@ export default class FilterOptionsComponent extends Component {
     });
   }
 
+  getLocationRadiusFilterNode () {
+    const selectedOption = this.config.options.find(o => o.selected);
+    if (!selectedOption) {
+      return FilterNodeFactory.from();
```
Comment (tmeyer2115) [inline src/ui/components/filters/filteroptionscomponent.js:392]: This is a minor nit, but do we want to return you anything if no option was selected?

File: src/ui/sass/modules/_ResultsHeader.scss [hunk 17]
```diff
@@ -28,11 +28,11 @@ $results-header-filters-line-height: var(--yxt-line-height-md) !default;
   --yxt-results-header-filters-color: #{$results-header-filters-color};
   --yxt-results-header-filters-line-height: #{$results-header-filters-line-height};
 }
-
 .yxt-ResultsHeader {
   padding: calc(var(--yxt-results-header-spacing) / 2) var(--yxt-results-header-spacing);
   display: flex;
   align-items: baseline;
+  height: 20px;
```
Comment (tmeyer2115) [inline src/ui/sass/modules/_ResultsHeader.scss]: Are these changes part of another PR?

File: src/core/filters/filterregistry.js (hunk 15 above)
Comment (oshi97) [reply src/core/filters/filterregistry.js]: got it

File: src/ui/components/filters/filteroptionscomponent.js (hunk 16 above)
Comment (oshi97) [reply src/ui/components/filters/filteroptionscomponent.js:392]: I thought maybe it was simpler to just return an empty filternode? alternatively I could add a clearLocationRadiusFilterNode() method

File: src/ui/sass/modules/_ResultsHeader.scss (hunk 17 above)
Comment (oshi97) [reply src/ui/sass/modules/_ResultsHeader.scss]: rebased!

File: src/core/models/filter.js (hunk 14 above)
Comment (oshi97) [reply src/core/models/filter.js]: removed

File: src/core/filters/filterregistry.js (hunk 13 above)
Comment (oshi97) [reply src/core/filters/filterregistry.js]: added a todo

Comment (tmeyer2115) [reply src/core/filters/filterregistry.js]: Sorry, I should have clarified. I think we should add the getFilterNodeByKey in this PR and not include this locationRadiusPayload method. We can remove the other payload methods later, but I don't think we should add another in this PR.

File: src/core/models/filter.js [hunk 18]
```diff
@@ -192,6 +192,13 @@ export default class Filter {
...
```
Comment (tmeyer2115) [inline src/core/models/filter.js]: Can we remove this static method?

Comment (oshi97) [reply src/core/models/filter.js]: sorry I must have mouse slipped or something

File: src/core/filters/filterregistry.js (hunk 13 above)
Comment (oshi97) [reply src/core/filters/filterregistry.js]: I see, removed locationRadiusPayload method and added getFilterNodeByKey

File: src/core/core.js [hunk 19]
```diff
@@ -154,7 +156,7 @@ export default class Core {
         queryTrigger: this.globalStorage.getState('queryTrigger'),
         sessionTrackingEnabled: this.globalStorage.getState(StorageKeys.SESSIONS_OPT_IN),
         sortBys: this.globalStorage.getState(StorageKeys.SORT_BYS),
-        locationRadius: this.filterRegistry.getLocationRadiusPayload()
+        locationRadius: locationRadiusFilterNode ? locationRadiusFilterNode.getFilter().value : null
```
Comment (tmeyer2115) [inline src/core/core.js:159]: Just to confirm, is locationRadius: null the same thing as leaving out locationRadius on the back-end?

**Review by tmeyer2115 — APPROVED** [review]
LGTM, just had the one question!

File: src/core/core.js (hunk 19 above)
Comment (oshi97) [reply src/core/core.js:159]: yup, we have a data cleaning step that removes all null and undefined params
